### ISSUE TITLE
Fash hash threat model might be optimistic

### ISSUE THREAD

**Issue by Iiridayn** [dropbox/zxcvbn#272]
Fast hash threat model assumes 2^58.13 hashes per year ([math](https://www.wolframalpha.com/input/?i=log2%28%2810+billion+per+second+in+years%29+*+year%29)). In [Towards Reliable Storage of 56-bit secrets in human memory](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/SecretsInHumanMemory_Extended.pdf) the researchers estimate an attacker could hit 2^70 SHA256 hashes per year at a cost of US$1M in 2013 (pg 2) based off of bitcoin hash rates, which seems a very reasonable metric.

To match that alone, the 10 billion / second would need to upgrade to 40 trillion / second [(math)](https://www.wolframalpha.com/input/?i=log2%28%2840+trillion+per+second+in+years%29+*+year%29) (US style affixes).

I've not analyzed the slow hash times to see if I'd agree with them, but I'd gather they can be safely derived from the fast hash times, assuming bcrypt et al - per the above paper, adding 14-16 bits might be safe. Actually, using the original 2^58 per year, we have over 100k / second with 16 bits added ([math](https://www.wolframalpha.com/input/?i=2%5E%2858-16%29+per+year+in+seconds)) or (rounding up for pessimism) about 1B / second for the 2013 2^70 / year figures ([math](https://www.wolframalpha.com/input/?i=2%5E%2870-16%29+per+year+in+seconds)).

On the other hand, your estimates might be entirely reasonable depending on the attacker. What are the attacker's presumed capabilities?

In an unrelated note, I think 100 / second might be equally feasible for an unthrottled online attack, given my experiences with negligent server administrators and poor password storage. It seems feasible with increasing bandwidth and server speeds that the unthrottled online attack also might change, though perhaps more slowly than offline attacks.

Contrariwise, the throttled online attack might well be much reduced from 100/hour - in the case of a debit card PIN for example, the limit is traditionally just 3 attempts before non-temporal lockout. Often systems have a much lower lockout throttle than 100/hour (though, being pessimistic is better than optimistic).

Current bitcoin hashrates appear to have peaked at time of this bug report at around 110,000,000 TH/s - or over 100 million trillion hashes, 1.10 * 10^20. That was on the 23nd of October ([one source](https://bitinfocharts.com/comparison/bitcoin-hashrate.html)). [Same site ](https://bitinfocharts.com/comparison/bitcoin-mining_profitability.html) estimates an income of 0.149 per TH for the same date, for an approximate cost in late 2019 of $54.38 USD per TH/s ([math](https://www.wolframalpha.com/input/?i=%28110%2C000%2C000+terahash+per+second%29+*+%28%240.149+USD+per+day+per+%28terahash+per+second%29%29+in+years+%2F+110000000)).

If an attacker has $1,000,000 to spend in 2019 for a year of cracking, they can crack approximately 5.8 * 10^23 hashes that year ([math](https://www.wolframalpha.com/input/?i=%241000000+%2F+%28%28%240.149+USD+per+day+per+%28terahash+per+second%29%29+in+years%29)) which is about (wolfram alpha finally gave up on my odd equations)   79 bits ([math](https://www.wolframalpha.com/input/?i=log2%285.799+*+10%5E23%29)) of entropy in a year.

Possibly the attacker wants to spread the cracking around to multiple accounts, or has less than $1,000,000 (or more than $1,000,000) to spend. If we give the attacker 100 times more money to attack 10,000 accounts (both somewhat reasonable pessimistic assumptions) the numbers thankfully cancel out reasonably leaving about 72.3 bits of true entropy cracked for each of 10,000 accounts in a year with a USD $100,000,000 system. (Can add the log2 of the ratio between money multiplies and accounts multiplier - ie, log2(100 / 10000) for $100M USD and 10,000 accounts to play with money/spread easily - ie, a $10k rig targeting one account for 1 year would also hit 72.3 bits of entropy as the ratio is the same).

Anyway - all that to say that I think 58 bits cracked per year per account is too optimistic. Around 40TH/s seems more reasonable to me, or if really needing a strict 10, 10 TH/s results in about 68 bits in a year, or about $500 per user per 2019 year.

**Comment by Iiridayn** [comment]
TL;DR: 40 TH/s for the fast offline hash, about 600 M/s for the slow offline hash, and 100 / second for the unthrottled online attack.
